src/data/nat/count.lean

```diff
@@ -0,0 +1,543 @@
+/-
+Copyright (c) 2021 Vladimir Goryachev. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Vladimir Goryachev, Kyle Miller, Scott Morrison, Eric Rodriguez
+-/
+import data.list.basic
+import data.nat.prime
+import order.order_iso_nat
+import set_theory.fincard
+
+/-!
+# Counting on ℕ
+
+This files defines ways to get basic properties of a predicate on ℕ, such as "how many numbers
+under `k` satisfy the predicate" and "what is the `n`th number that satisifies this predicate".
+We define these as two functions, `count` and `nth`, that answer these questions, and prove
+the expected theorems about them.
+
+## Main definitions
+
+* `count p n`: The number of naturals `k < n` such that `p k`.
+* `nth p n`: The `n`-th natural `k` (zero-indexed) such that `p k`. If there is no
+  such natural (that is, `p` is true for at most `n` naturals), then `nth p n = 0`.
+
+## Main results
+
+* `set.infinite.order_iso_nat`: An infinite set of natural numbers is order-isomorphic to the
+  natural numbers.
+-/
+
+lemma list.range_add (a : ℕ) :
+  ∀ b, list.range (a + b) = list.range a ++ (list.range b).map (λ x, a + x)
+| 0 := by rw [add_zero, list.range_zero, list.map_nil, list.append_nil]
+| (b + 1) := by rw [nat.add_succ, list.range_succ, list.range_add b, list.range_succ,
+  list.map_append, list.map_singleton, list.append_assoc]
+
+lemma multiset.range_add (a b : ℕ) :
+  multiset.range (a + b) = multiset.range a ∪ (multiset.range b).map (λ x, a + x) :=
+begin
+  rw [multiset.range, list.range_add, ←multiset.coe_add, ←multiset.coe_map],
+  refine multiset.add_eq_union_iff_disjoint.2 (λ x hxa hxb, _),
+  rw [multiset.mem_coe, list.mem_range] at hxa,
+  obtain ⟨c, _, rfl⟩ := multiset.mem_map.1 hxb,
+  exact (self_le_add_right _ _).not_lt hxa,
+end
+
+lemma finset.range_add (a b : ℕ) :
+  finset.range (a + b) = finset.range a ∪ (finset.range b).image (λ x, a + x) :=
+begin
+  rw [←finset.val_inj, finset.range, finset.union_val,
+    finset.image_val_of_inj_on ((add_right_injective a).inj_on _)],
+  exact multiset.range_add a b,
+end
+
+lemma galois_connection.lt_iff_lt {α β : Type*} [linear_order α] [linear_order β] {l : α → β}
+  {u : β → α} (gc : galois_connection l u) {a : α} {b : β} :
+  b < l a ↔ u b < a :=
+lt_iff_lt_of_le_iff_le (gc a b)
+
+open finset
+
+namespace nat
+variable (p : ℕ → Prop)
+
+section count
+variable [decidable_pred p]
+
+/-- Count the number of naturals `k < n` satisfying `p k`. -/
+def count (n : ℕ) : ℕ := ((list.range n).filter p).length
+
+@[simp] lemma count_zero : count p 0 = 0 :=
+by rw [count, list.range_zero, list.filter_nil, list.length]
+
+/-- A fintype instance for the set relevant to `nat.count`. Locally an instance in locale `count` -/
+def count_set.fintype (n : ℕ) : fintype {i | i < n ∧ p i} :=
```

Comment:
This would feel more natural to me as
```suggestion
def count_set.fintype (n : ℕ) : fintype {i // i < n ∧ p i} :=
```
which avoids a weird set coercion.